File: 0572-SubtreeOfAnotherTree/Solution.java
import java.util.*;

public class Solution {
    
    public boolean isSubtree(TreeNode s, TreeNode t) {
        if (s == null) { // t is non-null
            return false;
        } else {
            return isSameTree(s, t)
                || isSubtree(s.left, t)
                || isSubtree(s.right, t);
        }
    }

    private boolean isSameTree(TreeNode s, TreeNode t) {
        if (s == null && t == null) {
            return true;
        } else if (s == null || t == null) {
            return false;
        } else {
            return s.val == t.val
                && isSameTree(s.left, t.left)
                && isSameTree(s.right, t.right);
        }
    }

    public static void main(String[] args) {
        System.out.println();
    }

}

File: 0572-SubtreeOfAnotherTree/solution.py
class Solution:

    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        if not s:  # t is never None
            return False
        else:
            return (self.isSameTree(s, t)
                    or self.isSubtree(s.left, t)
                    or self.isSubtree(s.right, t))

    def isSameTree(self, s: TreeNode, t: TreeNode) -> bool:
        if not s and not t:
            return True
        elif not s or not t:
            return False
        else:
            return (s.val == t.val
                    and self.isSameTree(s.left, t.left)
                    and self.isSameTree(s.right, t.right))


if __name__ == "__main__":
    s = Solution()

